src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,98 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizer;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.CostModel;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.LogicalToPhysicalPlanConverter;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RewritingRule;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleApplication;
+
+public class IterativeImprovementBasedQueryOptimizer implements QueryOptimizer {
+	
+	protected final QueryOptimizationContext context;
+	LogicalToPhysicalPlanConverter converter;
+	
+	public IterativeImprovementBasedQueryOptimizer (final QueryOptimizationContext ctxt) {
+		assert ctxt != null;
+		context = ctxt;
+	}
+	
+	
+	@Override
+	public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimizationException {
+
+		converter = context.getLogicalToPhysicalPlanConverter();
+		PhysicalPlan optimalPlan = converter.convert(initialPlan, false);
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel();
+		CompletableFuture<Double> futureCost = costModel.initiateCostEstimation(optimalPlan);
+		
+		// Create rule instances.
+
+		// Get the cost.
+		Double localMinimum;
+		try {
+			localMinimum = futureCost.get();
+		} catch(final ExecutionException e) {
+			throw new QueryOptimizationException("CompletableFuture throws ExecutionException!", e);
+		} catch(final InterruptedException e) {
+			throw new QueryOptimizationException("CompletableFuture throws InterruptedException!", e);		
+		}
+		Double optimalCost, currentCost = localMinimum;
+		// currentCost = cost of the plan whose neighbours is being looked at.
+		// localMinimum = best cost among all neighbours
+		// optimalCost = best possible cost out of everything that the algorithm has found.
+		
+		// Optimization takes place here.
+		// Given the one starting state, find a local minimum.
+		// Firstly, generate neighbours. (Come to think of it, should I use the British or American spelling of neighbours?)
+		RewritingRule rwRule = null; // Eclipse suggested initiating this by adding the "= null". This doesn't seem quite right...
+		List<PhysicalPlan> neighbours = getNeighbours(optimalPlan,rwRule);
+		
+		List<CompletableFuture<Double>> futureNeighbourCosts = null; // Again, null suggested by Eclipse. Is this right?

```

Comment:
No, in this case you certainly do not want to assign `null` to this variable. Instead, you want to create an actual list and assign that one. To this end, you can do:
```suggestion
		List<CompletableFuture<Double>> futureNeighbourCosts = new ArrayList<>();
```